pkg/banneduser/banneduser_test.go

```diff
@@ -0,0 +1,188 @@
+			wantError:    false,
+			wantErrorMsg: "",
+			expectedBannedUser: &toolchainv1alpha1.BannedUser{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: userSignup3.Namespace,
+					Name:      fmt.Sprintf("banneduser-%s", userSignup3UserEmailHashLabelKey),
+					Labels: map[string]string{
+						toolchainv1alpha1.BannedUserEmailHashLabelKey: userSignup3UserEmailHashLabelKey,
+						bannedByLabel: "admin",
+						toolchainv1alpha1.UserSignupUserPhoneHashLabelKey: userSignup3PhoneHashLabelKey,
+					},
+				},
+				Spec: toolchainv1alpha1.BannedUserSpec{
+					Email: userSignup3.Spec.IdentityClaims.Email,
+				},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := NewBannedUser(tt.userSignup, tt.bannedBy)
+
+			if tt.wantError {
+				require.Error(t, err)
+				assert.Equal(t, tt.wantErrorMsg, err.Error())
+				require.Nil(t, got)
+			} else {
+				require.NoError(t, err)
+				require.NotNil(t, got)
+
+				assert.Equal(t, tt.expectedBannedUser.ObjectMeta.Namespace, got.ObjectMeta.Namespace)
+				assert.Equal(t, tt.expectedBannedUser.ObjectMeta.Name, got.ObjectMeta.Name)
+				assert.Equal(t, tt.expectedBannedUser.Spec.Email, got.Spec.Email)
+
+				if tt.expectedBannedUser != nil && !compareMaps(tt.expectedBannedUser.Labels, got.Labels) {
+					t.Errorf("compareMaps(%v, %v) = false, expected = true", tt.expectedBannedUser.Labels, got.Labels)
+				}
+			}
+		})
+	}
+}
+
+func TestIsAlreadyBanned(t *testing.T) {
+	userSignup1 := commonsignup.NewUserSignup(commonsignup.WithName("johny"), commonsignup.WithEmail("johny@example.com"))
+	userSignup2 := commonsignup.NewUserSignup(commonsignup.WithName("bob"), commonsignup.WithEmail("bob@example.com"))
+	userSignup3 := commonsignup.NewUserSignup(commonsignup.WithName("oliver"), commonsignup.WithEmail("oliver@example.com"))
+	bannedUser1, err := NewBannedUser(userSignup1, "admin")
+	require.NoError(t, err)
+	bannedUser2, err := NewBannedUser(userSignup2, "admin")
+	require.NoError(t, err)
+	bannedUser3, err := NewBannedUser(userSignup3, "admin")
+	require.NoError(t, err)
+
+	mockT := test.NewMockT()
+	fakeClient := test.NewFakeClient(mockT, bannedUser1)
+	ctx := context.TODO()
+
+	tests := []struct {
+		name       string
+		toBan      *toolchainv1alpha1.BannedUser
+		wantResult bool
+		wantError  bool
+		fakeClient *test.FakeClient
+	}{
+		{
+			name:       "user is already banned",
+			toBan:      bannedUser1,
+			wantResult: true,
+			wantError:  false,
+			fakeClient: fakeClient,
+		},
+		{
+			name:       "user is not banned",
+			toBan:      bannedUser2,
+			wantResult: false,
+			wantError:  false,
+			fakeClient: fakeClient,
+		},
+		{
+			name:       "cannot list banned users because the client does have type v1alpha1.BannedUserList registered in the scheme",
+			toBan:      bannedUser3,
+			wantResult: false,
+			wantError:  true,
+			fakeClient: &test.FakeClient{Client: fake.NewClientBuilder().WithScheme(scheme.Scheme).Build(), T: t},
```

Comment:
That would work too. But just FYI. We usually replace the underlying client function to test error handling. Something like this:
```
fakeClient.MockList = func(ctx context.Context, list runtimeclient.ObjectList, opts ...runtimeclient.ListOption) error {
	return errors.New("Oopsie Woopsie")
}
```

Reply:
I really appreciate it! Thank you so much!